Docs: Styling edits for Deploying to Openshift guide

sheilamjones opened the pull request:
This PR aims to enhance the existing Deploying on OpenShift guide as follows:

- Complete iteration 1 of restructure of Deploying on OpenShift guide into Diataxis content types. Restructure updates include the following changes:
  - Introduce a new concept topic "deployment-to-openshift.adoc" to provide a overview of deployment to OpenShift concepts
  - Introduce 2 new "How-to" topics based on downstream Deploy to OpenShift guide content. The new topics are:
  -  Deploy Quarkus applications as an OpenShift Serverless service ("deploying-to-openshift-serverless.adoc")
  -  Deploy a Quarkus native application on OpenShift ("deploying-native-executable-openshift.adoc")

- Edit to enhance and style the existing guide as per the [Quarkus doc contributor guide](https://quarkus.io/guides/doc-reference)

- Implement corrections arising from Vale Linter checks

sheilamjones committed:
Docs: Styling edits for Deploying to Openshift guide

sheilamjones commented:
Hi @iocanel, @mkouba, @Ladicek,
Please see draft updates to the Deploying to OpenShift guide for which I would appreciate your review. Per our planning discussion a couple of weeks ago, I completed styling edits on the existing guide per style and Diataxis guidelines. I created a concept and new How To topics to backport corresponding Deploy to OpenShift content from downstream.  
Many thanks in advance,
Sheila

iocanel reviewed (changes requested):
Great work overall.

I started adding some comments to the files in the order they appeared in the review tab only to realize that lots of them are addressed by `deployment-to-openshift.adoc` that I assume is meant to be read first.

To be more specific the individual files felt like they are missing an overview that was presented in the `deployment-to-openshift.adoc`. That being said we can't be really sure how the user may land to each of this pages, so maybe a smaller overview (possibly pointing to `deployment-to-openshift.adoc` is needed.

Other than that added some minor comments here and there.

sheilamjones commented:
Hi @iocanel, many thanks for your review and feedback. Working to process these today. With regards to your commend about the overview and conceptual information, you are correct. The intention was to incorporate that information in the "deployment-to-openshift" conceptual topic and provide links from there to the individual How Tos. But, I could enhance the overviews of the individual How Tos per your suggestion and perhaps link to the larger concept topic at the beginning also.

sheilamjones commented on docs/src/main/asciidoc/deploying-native-executable-openshift.adoc:
```diff
@@ -0,0 +1,195 @@
++
+[source,bash]
+----
+quarkus.native.container-build=true
```
Updated to add "Optional" prefix to the introductory statement:

.. Optional: Set the container runtime:
+
[source,bash]
----
quarkus.native.container-build=true
----

sheilamjones commented on docs/src/main/asciidoc/deployment-to-openshift.adoc:
```diff
@@ -0,0 +1,83 @@
+
+* By using the Quarkus CLI
++
+You can issue the `quarkus deploy` command to enable the deployment of Quarkus applications to platforms, such as Kubernetes or OpenShift Container Platform.
```
Thank you. Updated and link also provided to https://quarkus.io/guides/cli-tooling guide for more detailed information.

sheilamjones commented on docs/src/main/asciidoc/deploying-native-executable-openshift.adoc:
```diff
@@ -0,0 +1,195 @@
+
+To log in by setting the `quarkus.kubernetes-client.api-server-url` configuration property and authenticating with the `quarkus.kubernetes-client.token`, or `quarkus.kubernetes-client.username` and `quarkus.kubernetes-client.password` respectively:
+
+:build-additional-parameters: -Dquarkus.kubernetes-client.api-server-url=myServerUrl -Dquarkus.kubernetes-client.token=myToken
```
Thank @iocanel. Can I clarify exactly what I need to change or advise what the command should for the QUARKUS_KUBERNETES_CLIENT_TOKEN environment variable alternative?

iocanel commented on docs/src/main/asciidoc/deploying-native-executable-openshift.adoc:
```diff
@@ -0,0 +1,218 @@
+
+Deploy native Quarkus applications to OpenShift Container Platform by using the Docker build strategy enabled by the Quarkus OpenShift extension.
+
+Quarkus supports the `quarkus-container-image-docker` extension for building container images.
```
I think that this paragraph is not accurate.
The docker build strategy enabled by the openshift extension has nothing to do with `quarkus-container-image-docker` it's a feature provided by `quarkus-container-image-openshift` that is automatically brought into the project when users add `quarkus-openshift` to their project.

sheilamjones replied:
Thank you. Deleted the line "Quarkus supports the `quarkus-container-image-docker` extension for building container images."

iocanel commented on docs/src/main/asciidoc/deploying-native-executable-openshift.adoc:
```diff
@@ -0,0 +1,218 @@
+Deploy native Quarkus applications to OpenShift Container Platform by using the Docker build strategy enabled by the Quarkus OpenShift extension.
+
+Quarkus supports the `quarkus-container-image-docker` extension for building container images.
+The extension uses the Docker binary and generated Dockerfiles in the `src/main/docker` directory to perform Docker builds.
```
The extension (meaning the quarkus-container-image-openshift) is not using the docker binary. It is uploading the native binary and the dockerfiles to openshift and openshift performs the build.

In fact this could all work without the user having docker installed. I am saying `could` because in some cases the `docker` binary may be used in order to create the native binary itself (as correctly described by the paragraph below).

sheilamjones replied:
Would the following update be correct instead:  "The `quarkus-container-image-openshift` extension uploads a native binary and Dockerfiles to OpenShift and OpenShift performs the Docker builds."

iocanel commented on docs/src/main/asciidoc/deployment-to-openshift.adoc:
```diff
@@ -0,0 +1,104 @@
+The artifacts include JAR files or a native executable.
+The container gets built inside the OpenShift Container Platform cluster and is provided as an image stream.
+
+== Build strategies supported by Quarkus
```
I think that this table is confusing and maybe we are better off without it.

sheilamjones replied:
Removed table.

iocanel reviewed (changes requested):
Did a second pass and resolved most of my previous comments that have been addressed.
Add a new one regarding the mention to `quarkus-container-image-docker` that I missed in my first pass.

sheilamjones commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -161,48 +193,120 @@ oc expose svc/greeting <3>
-In other words, you don't need to repeat the above steps.
+After this setup, the next time you build a container image, a deployment to OpenShift is triggered automatically so you do not need to repeat the above steps.
+
+=== Option: Configure OpenShift application manually by using the Docker build
```
Hi @iocanel, Thank you for your additional review and I updated the sections per your feedback. Following our meeting, I also updated the _Choose how you want to build and deploy your Quarkus application_ chapter to add the Docker build strategy as a subsection (content backported from downstream guide). Does this look okay to you?

sheilamjones commented:
Hi @iocanel, 
Thank you for your additional review and I updated the sections per your feedback. One outstanding section is the building using the docker build strategy, which I added following our meeting. 

@maxandersen, @tqvarnst, @cescoffier, @mkouba - I would appreciate your review of these updates also. The changes comprise a new concept topic, and the existing Openshift guide divided to comprise 3 how to guides (deploying to Openshift Container Platform, deploying to serverless, deploying native executable)

Many thanks,
Sheila

iocanel reviewed (approved):
I marked all my comments as resolved.
All? There is still one unresolved about the need of `docker` binary for the openshift docker build strategy. I confirm that the suggested change sounds good and once that's in all my concerns will be addressed.

cescoffier commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -49,70 +57,90 @@ When we added the OpenShift extension to the command line invocation above, the
-The generated resources use OpenShift's `DeploymentConfig` that is configured to automatically trigger a redeployment when a change in the `ImageStream` is noticed.
-In other words, any container image build after the initial deployment will automatically trigger redeployment, without the need to delete, update or re-apply the generated resources.
+A new container image build triggers a refresh of the Quarkus application in OpenShift, `-Dquarkus.container-image.build=true`, which might suffice in most situations.
+If you must update the OpenShift resources, you first delete the binding to create it again after a new deployment.
```
(First, the reader is probably not aware of what is service binding)

cescoffier commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -49,70 +57,90 @@ When we added the OpenShift extension to the command line invocation above, the
-Note that the service is not exposed to the outside world by default.
-So unless you've used the `quarkus.openshift.route.expose` config property to expose the created service automatically you'll need to expose the service manually.
+The service is not exposed to the outside world by default.
+Therefore, unless you have used the `quarkus.openshift.route.expose` configuration property to expose the created service automatically, you must expose the service manually.
```
That was mentioned before, I would remove it from above to mention is in a single place.

cescoffier commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -161,48 +193,120 @@ oc expose svc/greeting <3>
+=== Option: Configure OpenShift application manually by using the Docker build
 
-=== Non-S2I Builds
+Build and deploy your Quarkus application to OpenShift Container Platform by using the Docker build strategy.
```
I would just reference the other guide.

cescoffier commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -161,48 +193,120 @@ oc expose svc/greeting <3>
-To select which extension will be used for building the image:
+Select the extension you want to use to build the image, by choosing one of the following commands.
 
+* To use the `quarkus-container-image-docker` extenstion, run the following command:
```
```suggestion
* To use the `quarkus-container-image-docker` extension, run the following command:
```

cescoffier commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -161,48 +193,120 @@ oc expose svc/greeting <3>
 
 [[exposing_routes]]
-=== Exposing Routes
+=== Expose routes
```
REference content - should not be there

cescoffier commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -221,11 +327,12 @@ You don't necessarily need to add this property in the `application.properties`.
 ====
 
-==== Securing the Route resource
+==== Secure the Route resource
```
Reference content - should not be there

cescoffier commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -236,57 +343,57 @@ quarkus.openshift.route.tls.termination=passthrough
 ----
 
-=== Labels
+=== Add labels
```
Reference content - should not be there

cescoffier commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -236,57 +343,57 @@ quarkus.openshift.route.tls.termination=passthrough
 ----
 
-=== Annotations
+=== Add annotations
```
Reference content - should not be there

cescoffier commented on docs/src/main/asciidoc/deployment-to-openshift.adoc:
```diff
@@ -0,0 +1,98 @@
+
+OpenShift Container Platform is a Kubernetes-based platform for developing and running containerized applications.
+As an application developer, to deploy Quarkus applications OpenShift Container Platform, you first choose how you want to build the application, for example, by using the `quarkus-openshift` extension or the Quarkus CLI.
+Then, you choose how you want to build the container image.
```
I would say that a flow chart would be super nice!

cescoffier commented on docs/src/main/asciidoc/deployment-to-openshift.adoc:
```diff
@@ -0,0 +1,98 @@
+However, for compatibility with earlier Quarkus versions, the default build strategy is the OpenShift binary build.
+To select the OpenShift Container Platform Docker build strategy, use the `quarkus.openshift.build-strategy` property.
+====
+
```
Something is missing here.
The extensions mentioned above are creating the container, then you need to push it.

cescoffier commented on docs/src/main/asciidoc/deployment-to-openshift.adoc:
```diff
@@ -0,0 +1,98 @@
+To select the OpenShift Container Platform Docker build strategy, use the `quarkus.openshift.build-strategy` property.
+====
+
+If you use a non-OpenShift container image extension, an `ImageStream` is created that points to an external `dockerImageRepository`.
```
link image stream - openshift specific resource.

cescoffier commented:
Alright,

I think we need to split that PR

1. Update to the openshift guide should be done first in a separate PR
2. The new deployment of native to openshift in another separate PR
3. The knative one should be separated too
4. The concept guide (which require a bit of rework) should come last

sheilamjones commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -4,35 +4,43 @@ and pull requests should be submitted there:
-* OpenShift CLI (Optional, only required for manual deployment)
+* Access to an OpenShift Container Platform cluster and the latest compatible version of the `oc` tool installed (Minishift is a viable option)
+* Access to the link:https://docs.openshift.com/container-platform/4.9/cli_reference/openshift_cli/getting-started-cli.html[OpenShift CLI] (Optional, only required for manual deployment)
+* Mandrel or GraalVM installed and configured appropriately.
```
Replaced "Mandrel or GraalVM installed and configured appropriately" with "Docker or Podman installed and configured appropriately"

sheilamjones commented on docs/src/main/asciidoc/deploying-to-openshift.adoc:
```diff
@@ -4,35 +4,43 @@ and pull requests should be submitted there:
+
+Create a new project that includes the `quarkus-openshift` extension.
+To build and deploy your applications as a container image that runs inside your OpenShift Container Platform cluster, add the Quarkus OpenShift extension as a dependency to your project.
+Run the following command:
```
The first sentence was an introduction. I removed it. The following 2 sentences are the instruction and the command needed to add the extension as the dependency.

sheilamjones commented:
Update:
Closing this PR per advice to split this PR into smaller separate PRs that would cover each guide separately. Feedback and discussion from this PR will be transferred to the new PRs.
- PR: https://github.com/quarkusio/quarkus/pull/34903 - Deploying to OpenShift How To plus Reference guides
- PR: https://github.com/quarkusio/quarkus/pull/35268 - Deploy native applications to OpenShift
- PR: Deploy to Serverless and Deployment to OpenShift concept (to be added)
